src/commands/env/create/scratch.ts

```diff
@@ -0,0 +1,210 @@
+      summary: messages.getMessage('flags.duration-days.summary'),
+    }),
+    wait: Flags.duration({
+      unit: 'minutes',
+      defaultValue: 5,
+      min: 1,
+      char: 'w',
+      summary: messages.getMessage('flags.wait.summary'),
+    }),
+    'track-source': Flags.boolean({
+      default: true,
+      summary: messages.getMessage('flags.track-source.summary'),
+      hidden: true, // for future use when AuthInfo supports this field
+    }),
+    'api-version': Flags.orgApiVersion(),
+    'client-id': Flags.string({
+      char: 'i',
+      summary: messages.getMessage('flags.client-id.summary'),
+    }),
+  };
+
+  public async run(): Promise<ScratchCreateResponse> {
+    const lifecycle = Lifecycle.getInstance();
+
+    const { flags } = await this.parse(EnvCreateScratch);
+
+    const createCommandOptions: ScratchOrgRequest = {
+      clientSecret: flags['client-id'] ? await this.clientSecretPrompt() : undefined,
+      connectedAppConsumerKey: flags['client-id'],
+      durationDays: flags['duration-days'].days,
+      nonamespace: flags['no-namespace'],
+      noancestors: flags['no-ancestors'],
+      wait: flags.wait,
+      apiversion: flags['api-version'],
+      definitionjson: flags['definition-file']
+        ? await fs.promises.readFile(flags['definition-file'], 'utf-8')
+        : JSON.stringify({ edition: flags.edition }),
+    };
+
+    let lastStatus: string;
+    const baseUrl = flags['target-dev-hub'].getField(Org.Fields.INSTANCE_URL).toString();
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    lifecycle.on<ScratchOrgLifecycleEvent>(scratchOrgLifecycleEventName, async (data): Promise<void> => {
+      lastStatus = buildStatus(data, baseUrl);
+      this.spinner.status = lastStatus;
+    });
+
+    this.spinner.start('Scratch Org Process');
+    const { username, scratchOrgInfo, authFields, warnings } = await flags['target-dev-hub'].scratchOrgCreate(
+      createCommandOptions
+    );
+    this.spinner.stop(lastStatus);
+
+    await this.maybeSetAliasAndDefault(username, flags['set-default'], flags.alias);
+    await lifecycle.emit(
+      'postorgcreate',
+      Object.fromEntries(Object.entries(authFields).filter(([key]) => isHookField(key))) as PostOrgCreateHook
+    );
+
+    this.log(chalk.green(messages.getMessage('success')));
+    return { username, scratchOrgInfo, authFields, warnings, orgId: scratchOrgInfo.Id };
+  }
+
+  private async clientSecretPrompt(): Promise<string> {
+    const { secret } = await this.timedPrompt<{ secret: string }>([
+      {
+        name: 'secret',
+        message: messages.getMessage('prompt.secret'),
+        type: 'password',
+      },
+    ]);
+    return secret;
+  }
+
+  private async maybeSetAliasAndDefault(username: string, setDefault: boolean, alias?: string): Promise<void> {
+    if (!setDefault && !alias) {
+      return;
+    }
+    const authInfo = await AuthInfo.create({ username });
+    return authInfo.handleAliasAndDefaultSettings({
```

Comment:
@mshanemc I am going to log a bug that will modify this function to no-op if no changes are made. One shouldn't have to guard the call.

Reply:
That's a good thing to do, but this code was trying to avoid the cost of `AuthInfo.create` so I'm going to leave it.